editor/src/components/canvas/canvas-strategies/strategies/set-border-radius-strategy.tsx

```diff
@@ -0,0 +1,470 @@
+    default:
+      assertNever(corner)
+  }
+}
+
+interface BorderRadiusData<T> {
+  mode: BorderRadiusAdjustMode
+  borderRadius: BorderRadiusSides<T>
+}
+
+function borderRadiusFromElement(
+  element: ElementInstanceMetadata,
+): BorderRadiusData<CSSNumberWithRenderedValue> | null {
+  if (element == null || isLeft(element.element) || !isJSXElement(element.element.value)) {
+    return null
+  }
+
+  const renderedValueSides = element.specialSizeMeasurements.borderRadius
+  if (renderedValueSides == null) {
+    return null
+  }
+
+  const fromProps = borderRadiusFromProps(element.element.value.props)
+  const borderRadius = optionalMap(
+    (radius) => measurementFromBorderRadius(renderedValueSides, radius),
+    fromProps,
+  )
+
+  if (borderRadius == null) {
+    return null
+  }
+
+  const borderRadiusUpperLimit = maxBorderRadius(
+    size(element.specialSizeMeasurements.clientWidth, element.specialSizeMeasurements.clientHeight),
+  )
+
+  return {
+    mode: fromProps?.type === 'sides' ? 'individual' : 'all',
+    borderRadius: mapBorderRadiusSides(
+      (n) => adjustBorderRadius({ min: 0, max: borderRadiusUpperLimit }, n),
+      borderRadius,
+    ),
+  }
+}
+
+interface BorderRadiusFromProps {
+  type: 'sides' | 'borderRadius'
+  sides: BorderRadiusSides<CSSNumber>
+}
+
+function borderRadiusFromProps(props: JSXAttributes): BorderRadiusFromProps | null {
+  const simpleBorderRadius = simpleBorderRadiusFromProps(props)
+
+  const borderTopLeftRadius = defaultEither(
+    null,
+    getLayoutProperty('borderTopLeftRadius', right(props), ['style']),
+  )
+  const borderTopRightRadius = defaultEither(
+    null,
+    getLayoutProperty('borderTopRightRadius', right(props), ['style']),
+  )
+  const borderBottomLeftRadius = defaultEither(
+    null,
+    getLayoutProperty('borderBottomLeftRadius', right(props), ['style']),
+  )
+  const borderBottomRightRadius = defaultEither(
+    null,
+    getLayoutProperty('borderBottomRightRadius', right(props), ['style']),
+  )
+
+  if (
+    borderTopLeftRadius != null ||
+    borderTopRightRadius != null ||
+    borderBottomLeftRadius != null ||
+    borderBottomRightRadius != null
+  ) {
+    return {
+      type: 'sides',
+      sides: {
+        tl: borderTopLeftRadius ?? simpleBorderRadius?.tl ?? cssNumber(0),
+        tr: borderTopRightRadius ?? simpleBorderRadius?.tr ?? cssNumber(0),
+        bl: borderBottomLeftRadius ?? simpleBorderRadius?.bl ?? cssNumber(0),
+        br: borderBottomRightRadius ?? simpleBorderRadius?.br ?? cssNumber(0),
+      },
+    }
+  }
+
+  if (simpleBorderRadius != null) {
+    return {
+      type: 'borderRadius',
+      sides: simpleBorderRadius,
+    }
+  }
+
+  return null
+}
+
+function simpleBorderRadiusFromProps(props: JSXAttributes): BorderRadiusSides<CSSNumber> | null {
+  const borderRadius = getLayoutProperty('borderRadius', right(props), ['style'])
+  if (isRight(borderRadius) && borderRadius.value != null) {
+    return isLeft(borderRadius.value)
```

Comment:
An `Either` within an `Either`?

Reply:
seems like `CSSBorderRadius` is `type CSSBorderRadius = Either<CSSNumber, CSSBorderRadiusIndividual>`

I'll definitely check the opportunities for `foldEither`